test/unit/test_grouped_loader.py

```diff
@@ -0,0 +1,238 @@
+"""
+Tests for deepreg/dataset/loader/grouped_loader.py in
+pytest style
+"""
+from os.path import join
+
+import numpy as np
+import pytest
+
+from deepreg.dataset.loader.grouped_loader import GroupedDataLoader
+from deepreg.dataset.loader.h5_loader import H5FileLoader
+from deepreg.dataset.loader.nifti_loader import NiftiFileLoader
+
+FileLoaderDict = dict(nifti=NiftiFileLoader, h5=H5FileLoader)
+nifti_path = join("data", "test", "nifti", "grouped")
+h5_path = join("data", "test", "h5", "grouped")
+DataPaths = dict(nifti=nifti_path, h5=h5_path)
+
+
+def test_validate_data_files():
+    """
+    Test validate_data_files function looks for inconsistencies in the fixed/moving image and label lists.
+    If there is any issue it will raise an error, otherwise it returns None.
+    """
+    for key_file_loader, file_loader in FileLoaderDict.items():
+        for train_split in ["train", "test"]:
+            for labeled in ["True", "False", 1, 0]:
+                data_dir_path = join(DataPaths[key_file_loader], train_split)
+                image_shape = (64, 64, 60)
+                common_args = dict(
+                    file_loader=file_loader,
+                    labeled=labeled,
+                    sample_label="all",
+                    intra_group_prob=1,
+                    intra_group_option="forward",
+                    sample_image_in_group=False,
+                    seed=None if train_split == "train" else 0,
+                )
+
+                data_loader = GroupedDataLoader(
+                    data_dir_path=data_dir_path, image_shape=image_shape, **common_args
+                )
+
+                assert data_loader.validate_data_files() is None
+
+
+def test_num_intra_group_probs():
+    """
+    Test whether fewer than two groups throws an error for intra_group_prob < 1
```

Comment:
checking https://codecov.io/gh/DeepRegNet/DeepReg/pull/247/src/deepreg/dataset/loader/grouped_loader.py?before=deepreg/dataset/loader/grouped_loader.py, we will find that the code

```
raise ValueError(
                    "Mixing intra and inter groups is not supported when not sampling pairs."
                )
```

is not tested, we need to add one more test with `sample_image_in_group == False` and `intra_group_prob==0.5`.

I suggest to put this in the same test function `test_init` and use comments to describe different test cases, so that each function in the test file corresponds to a function in the code to be tested.

Reply:
Agreed, am wondering if better input parsing is required in group_loader to force consistency between `intra_group_prob`, `intra_group_option` and `sample_image_in_group` values.

Reply:
what do you mean input parsing? checking values in the init? I think it is definitely better, change as you say ;)